ext/cache/01_cache.js

```diff
@@ -0,0 +1,261 @@
+        }
+      } else if (typeof request === "string" || request instanceof URL) {
+        try {
+          r = new Request(request);
+        } catch (error) {
+          return PromiseReject(error);
+        }
+      }
+      return await core.opAsync("op_cache_delete", {
+        cacheId: this.#id,
+        requestUrl: r.url,
+      });
+    }
+
+    /** See https://w3c.github.io/ServiceWorker/#cache-matchall
+     *
+     * Note: the function is private as we don't want to expose
+     * this API to the public yet.
+     *
+     * The function will return an array of responses.
+     */
+    async #matchAll(request, options = {}) {
+      // Step 1.
+      let r = null;
+      // Step 2.
+      if (request instanceof Request) {
+        r = request;
+        if (request.method !== "GET" && !options["ignoreMethod"]) {
+          return PromiseResolve([]);
+        }
+      } else if (typeof request === "string" || request instanceof URL) {
+        try {
+          r = new Request(request);
+        } catch (error) {
+          return PromiseReject(error);
+        }
+      }
+
+      // Step 5.
+      const responses = [];
+      // Step 5.2
+      if (r === null) {
+        // Step 5.3
+        // Note: we have to return all responses in the cache when
+        // the request is null.
+        // We deviate from the spec here and return an empty array
+        // as we don't expose matchAll() API.
+        return responses;
+      } else {
+        const matchResult = await core.opAsync(
+          "op_cache_match",
+          {
+            cacheId: this.#id,
+            requestUrl: r.url,
+          },
+        );
+        if (matchResult) {
+          const [meta, responseBodyRid] = matchResult;
+          let body = null;
+          if (responseBodyRid !== null) {
+            body = new ReadableStream({
+              type: "bytes",
+              async pull(controller) {
+                try {
+                  // This is the largest possible size for a single packet on a TLS
+                  // stream.
+                  const chunk = new Uint8Array(16 * 1024 + 256);
+                  const read = await core.read(responseBodyRid, chunk);
+                  if (read > 0) {
+                    // We read some data. Enqueue it onto the stream.
+                    controller.enqueue(chunk.subarray(0, read));
+                  } else {
+                    // We have reached the end of the body, so we close the stream.
+                    core.close(responseBodyRid);
+                    controller.close();
+                  }
+                } catch (err) {
+                  // There was an error while reading a chunk of the body, so we
+                  // error.
+                  controller.error(err);
+                  controller.close();
+                  core.close(responseBodyRid);
+                }
+              },
+            });
+          }
+
+          const response = new Response(
```

Comment:
I think there are internal fastpaths for constructing `Response`/`Request`?

I'll check later.